core/src/test/java/kafka/testkit/TestKitNodes.java

```diff
@@ -59,87 +63,66 @@ public Builder setCombined(boolean combined) {
+                throw new RuntimeException("Invalid value for numDisksPerBroker");
             }
-            return this;
-        }
 
-        public TestKitNodes build() {
             String baseDirectory = TestUtils.tempDirectory().getAbsolutePath();
-            try {
-                if (clusterId == null) {
-                    clusterId = Uuid.randomUuid();
-                }
-                TreeMap<Integer, ControllerNode> controllerNodes = new TreeMap<>();
-                for (ControllerNode.Builder builder : controllerNodeBuilders.values()) {
-                    ControllerNode node = builder.
-                        build(baseDirectory, clusterId, brokerNodeBuilders.containsKey(builder.id()));
-                    if (controllerNodes.put(node.id(), node) != null) {
-                        throw new RuntimeException("Duplicate builder for controller " + node.id());
-                    }
-                }
-                TreeMap<Integer, BrokerNode> brokerNodes = new TreeMap<>();
-                for (BrokerNode.Builder builder : brokerNodeBuilders.values()) {
-                    BrokerNode node = builder.
-                        build(baseDirectory, clusterId, controllerNodeBuilders.containsKey(builder.id()));
-                    if (brokerNodes.put(node.id(), node) != null) {
-                        throw new RuntimeException("Duplicate builder for broker " + node.id());
-                    }
-                }
-                return new TestKitNodes(baseDirectory,
-                    clusterId,
-                    bootstrapMetadata,
-                    controllerNodes,
-                    brokerNodes);
-            } catch (Exception e) {
-                try {
-                    Files.delete(Paths.get(baseDirectory));
-                } catch (Exception x) {
-                    throw new RuntimeException("Failed to delete base directory " + baseDirectory, x);
-                }
-                throw e;
+            if (clusterId == null) {
+                clusterId = Uuid.randomUuid();
             }
+
+            Map<Integer, ControllerNode> controllerNodes = IntStream.range(startControllerId(), startControllerId() + numControllerNodes)
+                .boxed().collect(Collectors.toMap(Function.identity(), id -> ControllerNode.builder()
+                    .setId(id)
+                    .setBaseDirectory(baseDirectory)
+                    .setClusterId(clusterId)
+                    .setCombined(combined)
+                    .build()));
+
+            Map<Integer, BrokerNode> brokerNodes = IntStream.range(startBrokerId(), startBrokerId() + numBrokerNodes)
+                .boxed().collect(Collectors.toMap(Function.identity(), id -> BrokerNode.builder()
+                    .setId(id)
+                    .setNumLogDirectories(numDisksPerBroker)
+                    .setBaseDirectory(baseDirectory)
+                    .setClusterId(clusterId)
+                    .setCombined(combined)
```

Comment:
As the number of brokers can be different to number of controllers, the `combined` should be replaced by `controllerNodes.containsKey(id)`.